src/android/toga_android/widgets/base.py

```diff
@@ -1,18 +1,36 @@
 from ..libs.activity import MainActivity
 from ..libs.android_widgets import Gravity
 
+from rubicon.java.jni import java
+
 from toga.constants import CENTER, JUSTIFY, LEFT, RIGHT
 
 
+def _get_activity(_cache=[]):
+    """
+    Android Toga widgets need a reference to the current activity to pass it as `context` when creating
+    Android native widgets. This may be useful at any time, so we retain a global JNI ref.
+
+    :param _cache: List that is either empty or contains 1 item, the cached global JNI ref
+    """
+    if _cache:
+        return _cache[0]
+    # See MainActivity.onCreate() for initialization of .singletonThis:
+    # https://github.com/beeware/briefcase-android-gradle-template/blob/3.7/%7B%7B%20cookiecutter.formal_name%20%7D%7D/app/src/main/java/org/beeware/android/MainActivity.java
+    if MainActivity.singletonThis is None:
+        raise ValueError("Unable to find MainActivity.singletonThis from Python. This is typically set by "
+                         "org.beeware.android.MainActivity.onCreate().")
+    _cache.append(MainActivity(__jni__=java.NewGlobalRef(MainActivity.singletonThis)))
+    return _cache[0]
+
+
 class Widget:
     def __init__(self, interface):
         self.interface = interface
         self.interface._impl = self
         self._container = None
         self.native = None
-        # Capture a reference to the Java `MainActivity` instance, so that subclasses
-        # can pass it as `context` when creating native Android widgets.
-        self._native_activity = MainActivity.singletonThis
+        self._native_activity = _get_activity()
```

Comment:
I presume the underlying issue here is that the activity JNI reference needs to be a global reference so it will remain valid when the class is called from other threads? If so - I suspect this will fix some other bugs that have been reported recently (specifically, someone who started a *python* thread, and the code crashed as soon as it tried to invoke an Android API from that thread)

Reply:
> I presume the underlying issue here is that the activity JNI reference needs to be a global reference so it will remain valid when the class is called from other threads?

In the case of fixing the specific crash I noticed with DetailedList, not so much from _other threads_, but the _same thread_ but later on, in an Android (Java) widget `onClick` that triggers Python widget `onClick` code.

I'm not sure if they work across threads. [This page](https://latkin.org/blog/2016/02/01/jni-object-lifetimes-quick-reference/) says yes, but I can't find evidence for that in [Google's Android JNI docs.](https://developer.android.com/training/articles/perf-jni.html#local_and_global_references)

If someone starts a Python thread and tries to access Android APIs, if they're GUI APIs, they're probably going to run into trouble. ["Do not access the Android UI toolkit from outside the UI thread,](https://developer.android.com/guide/components/processes-and-threads#Threads) says the Android docs. I can't remember _how_ this policy is enforced (does the app abruptly crash? raise an exception?). Having said that, it's possible that this patch does fix the specific crashing issue that Toga user ran into; they just might have other problems now.